### PR TITLE
Feature/quic server arrange the demo server to test with chrome

### PR DESCRIPTION
**Pull request by jfclere**
Arrange the demo server:
1 - Use something like wait_for_activity
2 - Allow to send files
3 - Wait for listener and connection in read_from_ssl_ids()
4 - Use SSL_handle_events() and SSL_poll(SSL_POLL_FLAG_NO_HANDLE_EVENTS)

### CODE REVIEW

**Comment by jfclere** [comment]
I have rebased I am testing, I have a small improvement I will commit later today or tomorrow.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 1]
```diff
@@ -898,7 +1216,9 @@ static int run_quic_server(SSL_CTX *ctx, int fd)
         /*
          * Free the connection, then loop again, accepting another connection.
          */
-        SSL_free(conn);
+        close_all_ids(&h3ssl);
+        SSL_free(h3ssl.ssl_ids[1].s);
```
Comment (Sashan) [inline demos/http3/ossl-nghttp3-demo-server.c:1220]: ssl_ids[1] (connection) and ssl_ids[0] (listener) are special I would suggest to define macros LISTENER, CONNECTION to distinguish those two objects in ids array.

**Review by Sashan — CHANGES_REQUESTED** [review]
thank you for all the work and sorry for not coming back to you earlier. this is the first round with few suggestions. I plan to revisit your PR as your change is fairly large.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 2]
```diff
@@ -335,169 +448,301 @@ static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
...
                    (unsigned long long)item->revents, SSL_POLL_EVENT_W,
-                   (unsigned long long)ssl_ids[i].id);
+                   (unsigned long long)id);
         }
     }
-    return hassomething;
+    ret = hassomething;
+err:
+    if (id_to_remove != UINT64_MAX)
+        remove_id(id_to_remove, h3ssl);
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c]: I have fixed it.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 3]
```diff
@@ -335,169 +448,301 @@ static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
...
+static void handle_events_from_ids(struct h3ssl *h3ssl)
+{
+    struct ssl_id *ssl_ids = h3ssl->ssl_ids;
+    int i;
+
+    ssl_ids = h3ssl->ssl_ids;
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].s != NULL) {
+            if (SSL_handle_events(ssl_ids[i].s))
+                ERR_print_errors_fp(stderr);
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:684]: Correct we need only to  SSL_handle_events()  on the listener (for  SSL_POLL_EVENT_IC) and on the SSL connection, working on a fix.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 4]
```diff
@@ -13,12 +13,14 @@
 #include <openssl/quic.h>
 #include <openssl/ssl.h>
 #include <unistd.h>
+#include <sys/stat.h>
+#include <fcntl.h>
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:17]: fixed, but I can't test for the moment.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 5]
```diff
@@ -81,6 +122,40 @@ static void add_id(uint64_t id, SSL *ssl, struct h3ssl *h3ssl)
     printf("Oops too many streams to add!!!\n");
     exit(1);
 }
+static void add_id_at(uint64_t id, SSL *ssl, int at, struct h3ssl *h3ssl)
+{
+    struct ssl_id *ssl_ids;
+
+    ssl_ids = h3ssl->ssl_ids;
+    if (ssl_ids[at].s == NULL) {
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:137]: To support multiple connections we need to have a nghttp3_conn for each connection (SSL connection) and arrange the struct h3ssl  to keep track of that.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 6]
```diff
@@ -53,16 +72,38 @@ static void init_ids(struct h3ssl *h3ssl)
     ssl_ids = h3ssl->ssl_ids;
     for (i = 0; i < MAXSSL_IDS; i++) {
         ssl_ids[i].s = NULL;
-        ssl_ids[i].id = -1;
+        ssl_ids[i].id = UINT64_MAX;
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:82]: id = 0 is the first valid bidi value we use so we need  UINT64_MAX (-1 doesn't make sense for uint64_t).
may be I missed something...

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 7]
```diff
@@ -335,169 +448,301 @@ static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
...
-            ret = quic_server_h3streams(h3conn, h3ssl);
-            if (ret == -1) {
-                fprintf(stderr, "quic_server_h3streams failed!\n");
-                return -1;
+        if (item->revents & SSL_POLL_EVENT_OSU) {
+            /* at least one uni */
+            /* we have 4 streams from the client 2, 6 , 10 and 0 */
+            /* need 3 streams to the client */
+            printf("Create uni?\n");
+            processed_event = processed_event + SSL_POLL_EVENT_OSU;
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:574]: fixed.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 8]
```diff
@@ -98,6 +173,23 @@ static void set_id_status(uint64_t id, int status, struct h3ssl *h3ssl)
...
+
+    ssl_ids = h3ssl->ssl_ids;
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].id == id) {
+            printf("get_id_status: %llu to %d\n",
+                   (unsigned long long) ssl_ids[i].id, ssl_ids[i].status);
+            return ssl_ids[i].status;
+        }
+    }
+    printf("Oops can't set status, can't find stream!!!\n");
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:182]: fixing it in the next commit...

File: demos/http3/ossl-nghttp3-demo-server.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c]: fixed

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 9]
```diff
@@ -335,169 +448,301 @@ static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
...
+
+static void handle_events_from_ids(struct h3ssl *h3ssl)
+{
+    struct ssl_id *ssl_ids = h3ssl->ssl_ids;
+    int i;
+
+    ssl_ids = h3ssl->ssl_ids;
+    for (i = 0; i < MAXSSL_IDS; i++) {
+        if (ssl_ids[i].s != NULL) {
+            if (SSL_handle_events(ssl_ids[i].s))
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:671]: fixed.

File: demos/http3/ossl-nghttp3-demo-server.c (hunk 1 above)
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:1226]: I have arranged the logic.

**Comment by nhorman** [comment]
@jfclere Just FYI, I've been working on getting our CI setup to test our quic server code using the quic-interop-runner (https://github.com/quic-interop/quic-interop-runner), for which there is a chrome client instance we can use this server to test against.

In my branch (here)[https://github.com/nhorman/openssl/tree/http3-server], I've been experimenting with getting it all working, and as that branch currently stands, its passing the http3 test using our openssl container built with your server from this PR against the chrome client container.  So thats really good news.

It did require some minor modifications however.  The branch has several debug commits, but the important bits are in:
* commit 537e739c2ebf3a2c7d9c02a49cec56f1f266bd00, which modifies the content-type header to be application/octet-stream (required to indicate to chrome that the file requested should be downloaded, rather than displayed in a browser window), and the addition of a content-length header, to inform chrome as to the file size (optional I think, haven't tested without it yet).
* commit 68ade593fa66c4a791b755f16b468c7de4144406, which modifies the server to allow for the specification of a directory prefix from which to serve files, which the interop harness requires to be able to point the server container at the files to serve to the client.

unsure if you are interested/willing to incorporate these changes into your PR.  If not, I'm happy to do so when this is complete, but if you would be willing, it would really help us out in moving toward getting some robust interop testing going in our quic server side code.  Thanks for considering!

**Comment by jfclere** [comment]
@nhorman We probably need a little more  make_nv(&resp[num_nv++], "content-type"), to allow html/html to text/html.
I will merge and have a closer look on Monday, many thanks for the help.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 10]
```diff
@@ -335,169 +476,325 @@ static int read_from_ssl_ids(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
...
+
+    /* Process all the item we have polled */
+    item = NULL;
+    for (i = 0; i < numitem; i++) {
+        SSL *s;
+
+        if (!item)
+            item = items;
+        else
+            item++;
```
Comment (Sashan) [inline demos/http3/ossl-nghttp3-demo-server.c:492]: Perhaps we can simplify code here to something like this:
```
for (i = 0, item = items; i < numitem; i++, items++) {
```
line 485, and lines 489 - 492 can be deleted.

**Review by Sashan — CHANGES_REQUESTED** [review]
Thank you for addressing my earlier comments.
there are some more findings. the most important one is to address all warnings/errors on CI builds.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 11]
```diff
@@ -851,27 +1190,38 @@ static int run_quic_server(SSL_CTX *ctx, int fd)
...
+        hasnothing = 0;
         for (;;) {
-            int hasnothing;
 
-            if (waitsocket(fd, 60)) {
-                printf("hasnothing timeout\n");
-                /* XXX probably not always OK */
-                break;
+            if (!hasnothing) {
+                SSL *ssl = get_ids_connection(&h3ssl);
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c]: I have noted and fixed this one.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 12]
```diff
@@ -247,7 +385,7 @@ static int quic_server_h3streams(nghttp3_conn *h3conn, struct h3ssl *h3ssl)
         fflush(stderr);
         return -1;
     }
-    pstream = SSL_new_stream(ssl_ids[0].s, SSL_STREAM_FLAG_UNI);
+    pstream = SSL_new_stream(ssl_ids[1].s, SSL_STREAM_FLAG_UNI);
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:393]: fixed, thanks.

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 13]
```diff
@@ -146,6 +275,18 @@ static int on_recv_header(nghttp3_conn *conn, int64_t stream_id, int32_t token,
     fwrite(vvalue.base, vvalue.len, 1, stderr);
     fprintf(stderr, "\n");
 
+    if (token == NGHTTP3_QPACK_TOKEN__PATH) {
+        memcpy(h3ssl->url, vvalue.base, vvalue.len);
+        if (h3ssl->url[0] == '/') {
+            if (h3ssl->url[1] == '\0') {
+                strcpy(h3ssl->url, "index.html");
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:287]: strlcpy() / strncpy()

File: demos/http3/ossl-nghttp3-demo-server.c [hunk 14]
```diff
@@ -146,6 +275,18 @@ static int on_recv_header(nghttp3_conn *conn, int64_t stream_id, int32_t token,
...
+            } else {
+                memcpy(h3ssl->url, h3ssl->url + 1, vvalue.len - 1);
```
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c:289]: Oops, fixed.

File: demos/http3/ossl-nghttp3-demo-server.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (jfclere) [reply demos/http3/ossl-nghttp3-demo-server.c]: fixed.

**Review by Sashan — CHANGES_REQUESTED** [review]
Thanks for addressing my comments.

May I also ask you to address c-style reported here: https://github.com/openssl/openssl/actions/runs/12238133177/job/34146189386?pr=25859?

you can also run c-style checker locally:
`./util/check-format.pl demos/http3/ossl-nghttp3-demo-server.c `

thanks a lot.

**Review by Sashan — APPROVED** [review]
Thanks for fixing `SSL_free()` I don't insist on c-style. it can be sorted out in follow up PR.

**Comment by jfclere** [comment]
There is a small glitch with the PREFIX hanging, I prefer to make another PR later, is that OK?

**Comment by nhorman** [comment]
Can you elaborate on what you mean by "The PREFIX hanging"?

**Comment by jfclere** [comment]
The prefix is erased by init_ids() and filename seems incorrect ":path: /kdsavcauux" but "Can't get_file_length /www/kdavcauuux".
I have a fix  if needed.

**Comment by nhorman** [comment]
I had fixed that in my branch, if you look on the branch I provided, init_ids looks like this:
```
static void init_ids(struct h3ssl *h3ssl)
{
    struct ssl_id *ssl_ids;
    char *prior_fileprefix = h3ssl->fileprefix;
    int i;

    memset (h3ssl, 0, sizeof (struct h3ssl));
    h3ssl->fileprefix = prior_fileprefix;

    ssl_ids = h3ssl->ssl_ids;
    for (i = 0; i < MAXSSL_IDS; i++) {
        ssl_ids[i].id = UINT64_MAX;
    }
    h3ssl->id_bidi = UINT64_MAX;
}
```
fileprefix is saved and restored in that operation using the prior_fileprefix variable

**Comment by jfclere** [comment]
:path: /kdsavcauux" but "Can't get_file_length /www/kdavcauuux". is due to:
https://github.com/jfclere/openssl/blob/feature/quic-server/demos/http3/ossl-nghttp3-demo-server.c#L294
I only see it when running the interop tests, but:
memcpy(h3ssl->url, h3ssl->url + 1, len - 1);
is looking for problems.
I will prepare a new PR once the  interop tests are running smoothly on my box.
I have:
Test: http3 took 14.440209s, status: TestResult.SUCCEEDED
Run took 0:00:17.198377
```
+---------+---------+
|         | openssl |
+---------+---------+
| openssl |   ✓(3)  |
|         |         |
|         |   ✕()   |
+---------+---------+
...

```

**Comment by nhorman** [comment]
yeah, do it in another pr @jfclere

**Comment by nhorman** [comment]
merged to quic-server branch, thank you

**Pull request closed without merging**